Initiate API ideation (#9)
jackbsteinberg/std-toast: README.md

# std-toast
This document scopes out a web platform API for a 'toast' pop-up notification.

> ## Why?

Reviewer: Is this doc intended the be the toast explainer? If not, this section seems like it belongs in the explainer.

Reviewer 2: Yep, this is the explainer.

Modern web applications allow users to complete lots of actions per page,
which necessitates providing clear feedback for each action.
Toast notifications are commonly used to unobtrusively provide this feedback.

Many libraries in a variety of frameworks implement a version of toast
(see [research](./study-group/)),
but the web has no built-in API to address the use case.
By providing a toast API as part of the web platform's standard library,
the web becomes more competitive with other app development platforms,
and web application developers can spend less of their time and bytes
on implementing this pattern.

Toasts are also a deceptively-tricky pattern to get right.
They require special accessibility considerations,
and scenarios involving multiple toasts need special handling.
[Not all libraries account for these subtleties](./study-group/).
By providing a built-in toast control that fully handles these aspects,
we can level up the typical toast experience for both developers and users of the web.

> Finally,

Reviewer: This argument is not so strong, it's a bit like "we should have a new standard to replace all of the old standards". We need to make it clear why we are not just adding yet-another toast but we are adding something that intended to be a foundation for other toasts. Is it actually our goal that React, etc will base their toast on top of ours? Or are we just going to make sure that they can create a good toast by e.g. adding the top-layer API?

That said, I'm not even sure that that's our rationale here,

Reviewer 2: (I wrote this text.)

Yes, the goal is that other toasts should be based on ours, and e.g. if someone writes a file picker component that wants to use toasts, they can then use std-toast inside of it, instead of worrying about what toast library they need to depend on and whether that makes sense with the application.

Can you think of ways to make this argument stronger?

Reviewer 3: I don't understand your explanation in this comment Domenic. It sounds like you're saying: "by having a standard toast, then other standard APIs can use it." Is that right? If so, I don't really understand what this has to do with what the non-standard toasts do.

Reviewer 2: No, sorry, I was saying that other libraries can use it. Maybe this needs to be spelled out in more detail; it's suffering a bit here by trying to be a short-ish paragraph that can stand alongside the other two as a peer, instead of overshadowing them with a ton of detail.

The scenario I was considering was the following:

- The website author wants to use several complicated third-party controls. One is a file picker. One is a settings panel. One is an email newsletter signup widget.
- All of these different controls want to signal success or completion using toasts.
- All of these different controls are built by different library authors or component vendors.

In the current world, each library author/component vendor would have to pick a toast implementation, or write their own, in order to signal completion. This would lead to the page author likely having three different types of toasts on their page.

In the world with std-toast, each library author/component vendor would use std-toast. Then, when the page author composes the three different widgets onto their page, all the toasts work the same. They can all be styled using `std-toast { ... }` in CSS; they all coordinate to stack on top of each other together instead of overlapping; etc.

This is the "shared vocabulary" argument for standards, just on a UI component level. Like we unified the ecosystem of promises by adding promises to ES2015, or of buttons by adding `<button>` to HTML4, or of disclosure widgets by adding `<details>` to HTML5, the idea is to do the same for toasts.

Make sense?

Reviewer: > (I wrote this text.)
> 
> Yes, the goal is that other toasts should be based on ours, and e.g. if someone writes a file picker component that wants to use toasts, they can then use std-toast inside of it, instead of worrying about what toast library they need to depend on and whether that makes sense with the application.
> 
> Can you think of ways to make this argument stronger?

If it's intended that this is a lower layer for all toasts, then I don't think it's a matter of making it stronger, it's more making it clear that that's what it is. E.g. by actually showing that react or whatever can delete some code and use this with no losses.

Reviewer 2: Captured in https://github.com/jackbsteinberg/std-toast/issues/14, for the record

## Sample code

The standard toast can be used according to two different patterns.

The first defines a `<std-toast>` HTML element,
then shows it with configurations via a method on the element.
This can be used to declaratively predefine toasts the application will need,
and then show them inside the application logic.

```html
<script type="module">
import 'std:elements/toast';
</script>

<std-toast id="sample-toast" theme="info">
    Hello World!
</std-toast>
```

```js
document.querySelector('#sample-toast').show({
    duration: 3000
});
```

The second imports the `showToast()` function from the `"std:elements/toast"` module,
which takes in a message and some configurations and creates and shows a toast in the DOM.
This is more convenient for one-off toasts,
or for JavaScript-driven situations,
similar to how `alert()` is currently used.

```js
import { showToast } from 'std:elements/toast';

const toast = showToast("Hello World!", {
    theme: "info",
    duration: 3000
});
```

## Goals

Across popular toast implementations there are recurring patterns,
which the standard toast aims to accomplish natively.

- The component will be accessible by default;

>   native accessibility is is a strong priority for toasts,

Reviewer: ```suggestion
  native accessibility is a strong priority for toasts,
```

- Toast implementations are often shaped similarly,
  and a goal of the standard toast is to make it as easy as possible for developers
  to build and style toasts that conform to those common shapes.

    - Toasts almost always support a message, action, and close button.

    - Toasts often support an icon and a title.

- The positioning of the toast must be intuitive,
  so the standard toast will come with built-in support for common positions,
  as well as a sensible default.

- To balance ease of use with customization,
  the standard toast will support creating and showing with one JavaScript function,
  as well as writing a custom view with a `<std-toast>` element
  and showing that with a method.

- The standard toast will come with support for showing multiple toasts,
  either by stacking them in the view,
  or queueing them and displaying sequentially.

## Proposed API

The element is provided as a [built-in module](https://github.com/tc39/proposal-javascript-standard-library/blob/master/README.md),
named `"std:elements/toast"`.

### The `<std-toast>` element

#### Attributes

- [Global attributes](https://html.spec.whatwg.org/multipage/dom.html#global-attributes)

> - `open`: a boolean attribute, determining whether the toast is visible or not (according to the default styles). By default toasts are not shown.

Reviewer: Can this just be `visibility` defaulting to `hidden`? Not sure if it's OK to say that a speific element defaults to `hidden`.

Reviewer 2: The idea is that, like `<dialog>`, the default styling for `std-toast:not([open])` is `display: none`. I guess this could be made explicit in the default styling section, which is currently just a TODO.

Reviewer: If it's following a proven pattern, that's good.

> - `open`: a boolean attribute, determining whether the toast is visible or not (according to the default styles). By default toasts are not shown.

Reviewer: I presume we don't want these to be findable (with ctrl-f) when they are not disaplayed. Something should mention that.

Reviewer 2: Fair point; this would also be clearer if we included the default `display: none` styling.

- `theme`: one of `"default"`, ???, or ???, conveying the semantic priority of the toast, and influencing its styling (both default and user-provided)
    - TODO: decide on list of themes to natively support and create and style them.

> - `position`: one of `"top-left"`, `"top-center"`, `"top-right"`, `"middle-left"`, `"middle-center"`, `"middle-right"`, `"bottom-left"`, `"bottom-center"`, or `"bottom-right"`. The default (if the attribute is omitted or set to an invalid value) is `"bottom-right"`.

Reviewer: break the line at "The default".

Author: ```suggestion
- `position`: one of `"top-left"`, `"top-center"`, `"top-right"`, `"middle-left"`, `"middle-center"`, `"middle-right"`, `"bottom-left"`, `"bottom-center"`, or `"bottom-right"`.
The default (if the attribute is omitted or set to an invalid value) is `"bottom-right"`.
```

> - `position`: one of `"top-left"`, `"top-center"`, `"top-right"`, `"middle-left"`, `"middle-center"`, `"middle-right"`, `"bottom-left"`, `"bottom-center"`, or `"bottom-right"`. The default (if the attribute is omitted or set to an invalid value) is `"bottom-right"`.

Reviewer: Why bottom right? I feel like Android is bottom-center. If other devices are different then perhaps a system-default option would be good.

Author: I went back through and checked the libraries in the comparison table, 1 had no default positioning, 5 had the default set to the top (1 `top-center` and 4 `top-right`, and 6 had the default set to the bottom (2 `bottom-left`, 3 `bottom-center`, and 1 `bottom-right`). So the arbitrary decision `bottom-right` is probably not the correct one.

Author: I'll change it s.t. the default is open to further discussion, and I'll add a TODO for system-default

>     - TODO: this seems stylistic; should it be controlled via CSS instead?

Reviewer: Seems CSS to me too.

Author: It's definitely something that could be controlled by CSS, but the idea was to at least give built-in values, and potentially offer the user an easy way to position the toast without styling it.

Ex:
```js
showToast("Hello World!", {
    timeout: 3000,
    position: "bottom-center"
});
```

    - TODO: these are physical positions; should we use [logical ones](https://developer.mozilla.org/en-US/docs/Web/CSS/CSS_Logical_Properties/Basic_concepts) instead?

> - `closebutton`: a boolean attribute, determining whether an explicit close button is shown. By default toasts do not have a close button. TODO

Reviewer: break the line

Author: ```suggestion
- `closebutton`: a boolean attribute, determining whether an explicit close button is shown. 
By default toasts do not have a close button. TODO
```

> - `closebutton`: a boolean attribute, determining whether an explicit close button is shown. By default toasts do not have a close button. TODO

Reviewer: TODO?

Author: I had that in there because I wasn't going to do `closebutton` in the MVP, which I did for a few other properties. On second pass I think I'm just going to remove them, since they're implementation details we're pretty sure about.

Reviewer: But what is the TODO?

Author: TODO was in relation to how and where `closebutton` should show up, and what it should look like when it does. Added a more clarification.

> - `closebutton`: a boolean attribute, determining whether an explicit close button is shown. By default toasts do not have a close button. TODO

Reviewer: Where would such a button appear? In particular, is there an area of the toast that is somehow reserved for the system. For the toast to be usable as the foundation of other library's toasts, we need to allow total control (including drawing outside of the box) but we also need make it work well when the close button is requested. Maybe it just means that a toast without a native close button draws in the full area and one with one draws in an area that makes sense given the positioning of the button.

BTW, me asking this question is what prompted me to send the mail about considering moving this to a doc.

```html
<std-toast>Hello world!</std-toast>

<std-toast><p>Hello world!</p></std-toast>

<std-toast>
  <p>Hello world!</p>
  <button>Click me!</button>
</std-toast>

<std-toast>
  <p>Hello world!</p>
  <a href="https://example.com/" target="_blank">Click me!</button>
</std-toast>
```

These can equivalently be created via JavaScript:

```js
// TODO: write this once we figure out the JS API for supplying HTML messages,
// and for supplying customizable actions
```

More complex toasts,
that don't fit the above content model,
would require more custom handling on the part of the developer:

```html
<std-toast>
  <p>Hello world!</p>
  <p>A second paragraph?!</p>

  <button>Action 1</button>
  <button>Action 2</button>
</std-toast>
```

Such an unusual toast would still integrate with other toasts in terms of stacking and positioning behavior,
and some of the default styles that are inherited may be useful.
But the page author will need to handle the action clicks themselves (instead of using the `"actionclick"` event),
and will need to add additional styling to handle the extra contents.

TODO: when we have a prototype, link to/show an example of this in action.

#### Methods

> The `.show(options)` function shows a toast element,

Reviewer: no `hide`/`dismiss` method? If that's intentional, we should explain our reasons.

Reviewer 2: Good question. @jackbsteinberg, did this appear much in the libraries you surveyed?

Author: Yeah libraries usually had a `close` or `dismiss` function available, that's something I could add directly to the research if it would be a value add to have there.

Author: I'm adding this to the table of comparison in the study group.

Author: I added this to the API as well

Should those be a global setting?
Per container?
How do we deal with different toasts having different values?

#### Events

A `<std-toast>` element can fire the following events:

- `"open"`: the toast was shown
- `"close"`: the toast was closed, either explicitly by the user, or via the timeout.
  (Note: if animations were applied, the toast may not be entirely invisible at the time this event fires)
    - TODO: should we consider separate events for the start and end of any close animation?
      This seems hard to do correctly if the user customizes the animation, though.
- `"actionclick"`: the toast's call-to-action button or link was clicked, if one exists TODO

### `showToast(message, options)`

The `"std:elements/toast"` module also exports a convenience function,
`showToast()`,
which allows creating and showing toasts entirely from JavaScript.
Behind the scenes,
`showToast()` creates a `<std-toast>`,
sets it up using the given message and options,
inserts it as the last child of `<body>`,
and then adds the `open=""` attribute,
to make the toast visible.
Finally,
it returns the created `<std-toast>` element,
allowing further manipulation by script.

> TODO: will `showToast()` support putting the toast into a specified container,

Reviewer: How about returning the element so that script can immediately insert it somewhere else?

Reviewer 2: This seems to be covered above, by

> Finally,
> it returns the created `<std-toast>` element,
> allowing further manipulation by script.

Reviewer: OK, so do we need this TODO?

Author: We don't need the TODO, that behavior can be recreated by the developer, and we have a demo of how in the Common Patterns section. Removing the TODO now.

or only into `body`?

`message` is a string that will be inserted as a text node
(TODO: or as a `<p>` element?) into the created `<std-toast>`.

TODO: can `message` contain HTML?

`options` allows configuring both the attributes of the `<std-toast>`,
and the options for this particular showing of the toast.
Thus, the possible options are:

- `theme`, like the attribute
- `position`, like the attribute
- `closeButton`, like the attribute
- `duration`, like the `show()` option
- `multiple`, like the `show()` option
- `newestOnTop`, like the `show()` option

> TODO: how do we deal with actions with this API? Ideas:

Reviewer: By "deal with actions", you mean how does the author specify actions? Do we need to provide this? It seems pretty easy to provide your own buttons or to wrap this and provide actions.

If we do, can just take a list of `[['Button text', callback], ...], that will work fine with a `message` toast (maybe expose a numbered CSS `part` for each one and a `part` that selects them all). If the author is doing a HTML toast then they can easily do anything they want for buttons and I wonder what value we would be adding by including a way of doing actions in that case.

Reviewer 2: Right, this is about how the author specifies actions for the JavaScript API. I agree that for a HTML toast this is not an issue.

The author could always do it themselves, but if the point of the JS API is to be easy to use like other libraries, I think

```js
const toast = showToast("message");
const button = document.createElement("button");
button.textContent = "Click me";
button.onclick = doSomething;
toast.append(button);
```

is not super user-friendly.

Your proposal is, I think,

```js
const toast = showToast("message", {
  action: ['Click me', doSomething]
});
```

which is one reasonable nicer API shape. But there are other possibilities too, and we were thinking of leaving this as a TODO and discussing the possibilities in an issue. (E.g., what if you want an `<a>` instead of a `<button>`?)

Reviewer 3: It seems showToast is a convenience wrapper around std-toast and that's where the action is useful. But std-toast needs actionclick to service showToast. Am I reading that right? That layering seems a bit strange.

This feels like the sort of thing that should be an open issue that we inform primarily with partner feedback.

Reviewer: Yeah. Issue sounds good.

Author: #11

If instead libraries and components all use the standard toast,
the application developer can centrally style and coordinate them.

## Sample code

The standard toast can be used according to two different patterns.

The first defines a `<std-toast>` HTML element,
then shows it with configurations via a method on the element.
This can be used to declaratively predefine toasts the application will need,
and then show them inside the application logic.

```html
<script type="module">
import 'std:elements/toast';
</script>

<std-toast id="sample-toast" theme="info">
    Hello World!
</std-toast>
```

```js
document.querySelector('#sample-toast').show({
    duration: 3000
});
```

The second imports the `showToast()` function from the `"std:elements/toast"` module,
which takes in a message and some configurations and creates and shows a toast in the DOM.
This is more convenient for one-off toasts,
or for JavaScript-driven situations,
similar to how `alert()` is currently used.

```js
import { showToast } from 'std:elements/toast';

const toast = showToast("Hello World!", {
    theme: "info",
    duration: 3000
});
```

> ## Goals

Reviewer 2: We should add an explicit goal to provide the foundation around which other toast libraries can layer themselves, and provide more opinionated defaults, specific styles, or better framework integration. Maybe with a "TODO: link to an example".

Author: That seems like a good addition to me, in keeping with explaining the spirit of why we're building this.

But the page author will need to handle the action clicks themselves (instead of using the `"actionclick"` event),
and will need to add additional styling to handle the extra contents.

TODO: when we have a prototype, link to/show an example of this in action.

#### Methods

The `.show(options)` function shows a toast element,
by toggling its `open=""` attribute to true.
The `options` include:

- `duration`: how long to show the toast, in milliseconds. Defaults to ???
- `multiple`: TODO
- `newestOnTop`: TODO

TODO: how do `multiple` and `newestOnTop` work?
Should those be a global setting?
Per container?
How do we deal with different toasts having different values?

#### Events

A `<std-toast>` element can fire the following events:

- `"open"`: the toast was shown
- `"close"`: the toast was closed, either explicitly by the user, or via the timeout.
  (Note: if animations were applied, the toast may not be entirely invisible at the time this event fires)
    - TODO: should we consider separate events for the start and end of any close animation?
      This seems hard to do correctly if the user customizes the animation, though.

> - `"actionclick"`: the toast's call-to-action button or link was clicked, if one exists TODO

Reviewer 3: What is the benefit of including this as part of the toast API surface?

Reviewer 2: Fair question. I think an example would make this clearer. The idea would be that you could do a simple:

```js
const toast = showToast("Message", { action: "Click me" });
toast.onactionclick = () => { ... };
```

instead of something like

```js
const toast = showToast("Message", { action: "Click me" });
toast.querySelector("button:last-of-type").onclick = () => { ... };
```

which makes the assumption that the action is provided by the last button in the toast. (Which might need to be adjusted if there's a close button? Although maybe it'd be fine if the close button was in the shadow DOM??)

The reason we didn't write up this example yet is because the `{ action: "Click me" }` syntax is not settled (see above thread with @fergald).

We should probably mark this as more clearly a tentative idea.

Reviewer 3: Ah, I missed that there's an action argument to showToast. But, my same question applies to that. I see that you're already discussing exactly this with fergal above though, so won't duplicate here.

Author: #11

- `open`: a boolean attribute, determining whether the toast is visible or not (according to the default styles). By default toasts are not shown.
- `theme`: one of `"default"`, ???, or ???, conveying the semantic priority of the toast, and influencing its styling (both default and user-provided)
    - TODO: decide on list of themes to natively support and create and style them.
- `position`: one of `"top-left"`, `"top-center"`, `"top-right"`, `"middle-left"`, `"middle-center"`, `"middle-right"`, `"bottom-left"`, `"bottom-center"`, or `"bottom-right"`.
The default (if the attribute is omitted or set to an invalid value) is ???.
    - TODO: this seems stylistic; should it be controlled via CSS instead?
    - TODO: these are physical positions; 
      should we use [logical ones](https://developer.mozilla.org/en-US/docs/Web/CSS/CSS_Logical_Properties/Basic_concepts) instead?
    - TODO: should there be a system default for positions 
      (e.g. should Android toasts show up in a different location from macOS toasts)?
- `closebutton`: a boolean attribute, determining whether an explicit close button is shown. 
By default toasts do not have a close button.

> TODO: where should the `closebutton` show up relative to toast content and how customizable should that be?

Author: ```suggestion
    - TODO: where should the `closebutton` show up relative to toast content and how customizable should that be?
```

The default (if the attribute is omitted or set to an invalid value) is ???.
    - TODO([#13](https://github.com/jackbsteinberg/std-toast/issues/13)): should this positioning be an attribute or a style
- `closebutton`: a boolean attribute, determining whether an explicit close button is shown. 
By default toasts do not have a close button.
    - TODO: where should the `closebutton` show up relative to toast content,
    and how customizable should it be?

All attributes will be reflected as properties on the element's JavaScript interface.
For example:

```js
const toast = document.createElement('std-toast');
console.log(toast.open); // false
```

#### Contents

The `<std-toast>` element can be used as a generic container,
but will work best (in terms of default styling and events) if the developer conforms to the following content model:

- The first child node,
  typically either a text node or a container like `<p>`,
  provides a message.
- If more than one child node is present,
  then the last child element,
  which can be either an `<a>` or a `<button>`,
  provides a call to action.

> TODO([#17](https://github.com/jackbsteinberg/std-toast/issues/17)): what about title or icon? They should potentially also be accomodated, in a similar fashion.

Reviewer 2: I think we should create and link to a tracking issue for whether this setup is the best one, to give us a place to discuss and explore alternatives. (For example, using `<button slot="action">...</button>`, or `<toastaction><button>...</button></toastaction>`, or similar.)

Author: Addressed in #21

```html
<std-toast>
  <p>Hello world!</p>
  <p>A second paragraph?!</p>

  <button>Action 1</button>
  <button>Action 2</button>
</std-toast>
```

Such an unusual toast would still integrate with other toasts in terms of stacking and positioning behavior,
and some of the default styles that are inherited may be useful.
But the page author will need to handle the action clicks themselves (instead of using the `"actionclick"` event),
and will need to add additional styling to handle the extra contents.

TODO: when we have a prototype, link to/show an example of this in action.

#### Methods

- `.show(options)`: shows the toast element,
by toggling its `open=""` attribute to true.
The `options` include:
    - `duration`: how long to show the toast, in milliseconds. Defaults to ???
    - `multiple`: ???
    - `newestOnTop`: ???

TODO([#19](https://github.com/jackbsteinberg/std-toast/issues/19)): how do `multiple` and `newestOnTop` work?

- `.hide()`: hides the toast element,
by toggling its `open=""` attribute to false.
- `.toggle([state])`: toggles the toast element,
by hiding it if it's being shown and showing it if it's being hidden,

> or changing the `open=""` attribute to `state` if given.

Reviewer 2: ```suggestion
or alternately adding/removing the `open=""` attribute per `state` if `state` given.
```

#### Events

A `<std-toast>` element can fire the following events:

- `"show"`: the toast was shown

> - `"hide"`: the toast was hiddne, either explicitly by the user, or via the timeout.

Reviewer 2: ```suggestion
- `"hide"`: the toast was hidden, either explicitly by the user, or via the timeout.
```

    - TODO([#11](https://github.com/jackbsteinberg/std-toast/issues/11)): flesh out exactly how the action works and how this event is linked.

### `showToast(message, options)`

The `"std:elements/toast"` module also exports a convenience function,
`showToast()`,
which allows creating and showing toasts entirely from JavaScript.
Behind the scenes,
`showToast()` creates a `<std-toast>`,
sets it up using the given message and options,
inserts it as the last child of `<body>`,
and then adds the `open=""` attribute,
to make the toast visible.
Finally,
it returns the created `<std-toast>` element,
allowing further manipulation by script.

`message` is a string that will be inserted as a text node
(TODO: or as a `<p>` element?) into the created `<std-toast>`.

`options` allows configuring both the attributes of the `<std-toast>`,
and the options for this particular showing of the toast.
Thus, the possible options are:

- `theme`, like the attribute
- `position`, like the attribute
- `closeButton`, like the attribute
- `duration`, like the `show()` option
- `multiple`, like the `show()` option
- `newestOnTop`, like the `show()` option

> TODO([#11](https://github.com/jackbsteinberg/std-toast/issues/11)): How should actions be handled?

Reviewer 2: Also add a TODO linking to https://github.com/jackbsteinberg/std-toast/issues/15 questioning whether a showToast-toast should get auto-removed from the DOM after it hides.

### Default styles

The standard toast will come with these default styles,
which developers will be able to change to customize look and feel.

#### `std-toast:not([open])`

- `display: none`: this will keep the toast hidden from view and searching
when the `open` attribute is not present.

#### `std-toast([open])`

TODO: figure out some default styles, and state them here.
Additionally figure out default styles for themes.

### Appearance customization

> TODO: explain any API for appearance customization,

Reviewer 2: I think we should have a tracking issue for this. Similar to https://github.com/tkent-google/std-switch/issues/7, it just gives us a place to put thoughts as we develop the prototype and get an idea of what we're doing.

Author: Addressed in #20
